### PR TITLE
[CET-7367] Birds eye - group by and links

### PR DESCRIPTION
**Pull request by PavelZurek**
## Change description

[CET-7367](https://cortex1.atlassian.net/browse/CET-7367) [GetYourGuide] Birds Eye Report for team Scorecards in Backstage default to service grouping

Bird's eye view:
 - Reports filters determined from URL are now also persisted in URL after change
 - Fixes entity link to target the entity performance on the Scorecard instead of the entity detail
 - Changes the Group By option Service to Entity to avoid confusion in other entity tables
 - Removes Group By Team for Team scorecards

## Type of change

- [x] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)
- [ ] Improvement (improves codebase without changing functionality)

### Development

- [x] The changelog has been updated as appropriate
- [x] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

[CET-7367]: https://cortex1.atlassian.net/browse/CET-7367?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx [hunk 1]
```diff
@@ -13,25 +13,37 @@
...
-    (queryParams.get('groupBy') as GroupByOption) ?? GroupByOption.SERVICE,
+    (searchParams.get('groupBy') as GroupByOption) ?? GroupByOption.SERVICE,
   );
 
+  useEffect(() => {
+    const targetUrl = stringifyUrl({ url: location.pathname, query: {
+      groupBy: groupBy !== GroupByOption.SERVICE ? groupBy as string : undefined,
+    } });
+
+    if (`${location.pathname}${location.search}` !== targetUrl) {
```
Comment (igorRoog) [inline src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx:42]: Please add some explanation comment why we need to check that.

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 2]
```diff
@@ -40,13 +47,25 @@ export const HeatmapPage = () => {
...
   );
 
+  useEffect(() => {
+    const targetUrl = stringifyUrl({ url: location.pathname, query: {
+      scorecardId: selectedScorecardId ? `${selectedScorecardId}` : undefined,
+      groupBy: groupBy !== defaultFilters.groupBy ? groupBy as string : undefined,
+      headerType: headerType !== defaultFilters.headerType ? headerType as string : undefined,
+    } });
+
+    if (`${location.pathname}${location.search}` !== targetUrl) {
```
Comment (igorRoog) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:64]: same here

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 3]
```diff
@@ -56,6 +75,26 @@ export const HeatmapPage = () => {
     return buildUrl(queryParamsObj, location.pathname);
   }, [location, selectedScorecardId, groupBy, headerType]);
 
+  const scorecardsResult = useCortexApi(api => api.getScorecards());
+  const excludedGroupBys = useMemo(() => {
+    const teamBasedCardSelected = scorecardsResult.value?.some(
+      (scorecard) => scorecard.id === selectedScorecardId && scorecard.filter?.type === FilterType.TEAM_FILTER
```
Comment (igorRoog) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:81]: @FoodProduct Correct me if I'm wrong, but is it still possible to create a scorecard with filtertype == TEAM_FILTER? Or it is only for old scorecards and now we define them via `COMPOUND_FILTER`. I'm filtering for teams-related scorecards on Teams graph page, so probably you need to do it the same way in here:

https://github.com/cortexapps/brain-app/blob/9530ad64ba2c9782c1f301fc53ce0b73d82b5a7c/apps/%40cortexapps/brain-app/src/components/Pages/Graph/FlowGraph/GraphToolbar/HierarchyOptionsButton.tsx#L53-L56

Comment (igorRoog) [reply src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:81]: Right, so we need to look for team-based scorecards in the same way it is done on Graph page in brain-app.

Comment (PavelZurek) [reply src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx]: I had to copy paste the types for this as well, I simplified them as much as i could.
I also added check for the CQL filter. Could be maybe missing in the brain-app condition?

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 4]
```diff
@@ -56,6 +77,40 @@ export const HeatmapPage = () => {
     return buildUrl(queryParamsObj, location.pathname);
   }, [location, selectedScorecardId, groupBy, headerType]);
 
+  const scorecardsResult = useCortexApi(api => api.getScorecards());
+
+  const selectedScoreCard = useMemo(() => {
```
Comment (igorRoog) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:82]:
```suggestion
  const selectedScorecard = useMemo(() => {
```

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 5]
```diff
@@ -56,6 +77,40 @@ export const HeatmapPage = () => {
...
+    return scorecardsResult.value?.find((scorecard) => scorecard.id === selectedScorecardId)
+  }, [scorecardsResult, selectedScorecardId]);
+
+  const entityCategory = useMemo(() => {
+    return getEntityCategoryFromFilter(selectedScoreCard?.filter) ?? 'Entity';
+  }, [selectedScoreCard]);
+
+  const excludedGroupBys = useMemo(() => {
+    const teamBasedCardSelected = selectedScoreCard && (
```
Comment (igorRoog) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:91]: Maybe we could create util function for this check?

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 6]
```diff
@@ -56,6 +77,40 @@ export const HeatmapPage = () => {
...
+
+  const selectedScoreCard = useMemo(() => {
+    return scorecardsResult.value?.find((scorecard) => scorecard.id === selectedScorecardId)
+  }, [scorecardsResult, selectedScorecardId]);
+
+  const entityCategory = useMemo(() => {
+    return getEntityCategoryFromFilter(selectedScoreCard?.filter) ?? 'Entity';
+  }, [selectedScoreCard]);
+
+  const excludedGroupBys = useMemo(() => {
```
Comment (igorRoog) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:90]: It looks like those three useMemos are tightly coupled. What do you think about combining them into single useMemo, like:

```
  const { entityCategory, excludedGroupBys } = useMemo(() => {
    const selectedScorecard = scorecardsResult.value?.find((scorecard) => scorecard.id === selectedScorecardId)
    
    return {
      entityCategory: getEntityCategoryFromFilter(selectedScorecard?.filter) ?? 'Entity',
      excludedGroupBys: teamBasedCardSelected(selectedScoreCard) ? [GroupByOption.TEAM] : []
    };
  }, [scorecardsResult, selectedScorecardId]);
```

File: src/utils/ScorecardFilterUtils.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { CategoryFilter, FilterType, Scorecard } from "../api/types";
+
+export function isScorecardTeamBased(scorecard?: Scorecard): boolean {
+  return (
```
Comment (FoodProduct) [inline src/utils/ScorecardFilterUtils.ts:19]: Could we break up this logic into multiple `if` conditions to make it easier to read please? 🙏

File: src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx [hunk 8]
```diff
@@ -13,31 +13,42 @@
...
 
 export const AllScorecardsPage = () => {
   const location = useLocation();
-  const queryParams = new URLSearchParams(useLocation().search);
+  const navigate = useNavigate();
+  const [searchParams] = useSearchParams();
 
-  const [groupBy, setGroupBy] = useDropdown<GroupByOption>(
-    (queryParams.get('groupBy') as GroupByOption) ?? GroupByOption.SERVICE,
+  const [groupBy, _setGroupBy] = useState<GroupByOption|undefined>(
```
Comment (FoodProduct) [inline src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx:32]: Why are we prefixing these setters with `_`? This is typically used for unused or private variables/methods (by convention), which doesn't seem to be the case here.

Comment (PavelZurek) [reply src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx:32]: Because it should not be used, except in the `setGoupBy`.

Comment (FoodProduct) [reply src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx:32]: So... it is used 😅

**Review by FoodProduct — APPROVED** [review]
Did not test e2e but code LGTM

**Review by igorRoog — APPROVED** [review]
E2E tests done. LGTM.

**Pull request merged**